C#/Botkit/botbuilder-slack-adapter/SlackMessageTypeMiddleware.cs

```diff
@@ -14,9 +42,65 @@ public class SlackMessageTypeMiddleware : MiddlewareSet
         /// </summary>
         /// <param name="context"></param>
         /// <param name=""></param>
-        public async void OnTurn(TurnContext context, Func<object>)
+        public async void OnTurn(TurnContext context, Func<Task<object>> next)
         {
+            if (context.Activity.Type == "message" && context.Activity.ChannelData != null)
+            {
+                var adapter = context.Adapter as SlackAdapter;
+
+                string bot_user_id = await adapter.GetBotUserByTeam(context.Activity); // TODO const
+                var mentionSyntax = "<@" + bot_user_id + "(\\|.*?)?>";
+                var mention = new Regex(mentionSyntax, RegexOptions.IgnoreCase);
+                var direct_mention = new Regex('^' + mentionSyntax, RegexOptions.IgnoreCase);
+
+                // is this a DM, a mention, or just ambient messages passing through?
+                if ((context.Activity.ChannelData as dynamic)?.channel_type == "im")
+                {
+                    (context.Activity.ChannelData as dynamic).botkitEventType = "direct_message";
+
+                    // strip any potential leading @mention
+                    Regex.Replace(
+                        Regex.Replace(
+                            Regex.Replace(
+                                Regex.Replace(context.Activity.Text, direct_mention.ToString(), ""), 
+                                @"/ ^\s +/", ""), 
+                            @"/ ^:\s +/", ""), 
+                        @"/ ^\s +/", "");
+                }
+                else if (!string.IsNullOrEmpty(bot_user_id) && !string.IsNullOrEmpty(context.Activity.Text) && context.Activity.Text.Equals(direct_mention))
+                {
+                    (context.Activity.ChannelData as dynamic).botkitEventType = "direct_mention";
+
+                    // strip the @mention
+                    Regex.Replace(
+                        Regex.Replace(
+                            Regex.Replace(
+                                Regex.Replace(context.Activity.Text, direct_mention.ToString(), ""),
+                                @"/ ^\s +/", ""),
+                            @"/ ^:\s +/", ""),
+                        @"/ ^\s +/", "");
+                }
+                else if (!string.IsNullOrEmpty(bot_user_id) && string.IsNullOrEmpty(context.Activity.Text) && context.Activity.Text.Equals(mention))
+                {
+                    (context.Activity.ChannelData as dynamic).botkitEventType = "mention";
+                }
+                else
```

Comment:
As this `else` doesn't have code lines, I'll assume it's not necessary. Remove it